File: frontend/src/services/analytics.ts
import { Budget, Category, MonthlyData, MonthlySummary, Movement, Recommendation } from '../types/finance';
import { perfMeasureAsync } from '../utils/performance';
import { decorateMovements, movementsService } from './movements';

export interface FinancialDataset {
  movements: Movement[];
  latestMovements: Movement[];
  categories: Category[];
  budget: Budget | null;
  currentBalance: number;
}

export interface FinancialAnalysis extends MonthlyData {
  currentBalance: number;
  trends: { year: number; month: number; label: string; income: number; expense: number; saving: number; balance: number }[];
  latestMovements: Movement[];
  documentsReadMaximum: number;
}

const money = (value: number) => new Intl.NumberFormat('es-CO', { style: 'currency', currency: 'COP', maximumFractionDigits: 0 }).format(value);
const periodKey = (year: number, month: number) => `${year}-${String(month).padStart(2, '0')}`;
const monthStart = (year: number, month: number) => `${periodKey(year, month)}-01`;
const monthEnd = (year: number, month: number) => `${periodKey(year, month)}-31`;

function shiftMonth(year: number, month: number, delta: number) {
  const date = new Date(Date.UTC(year, month - 1 + delta, 1));
  return { year: date.getUTCFullYear(), month: date.getUTCMonth() + 1 };
}

function monthlyBalanceDelta(item: Movement) {
  if (item.type === 'income') return item.amount;
  if (item.type === 'expense' || item.type === 'purchase') return -item.amount;
  if (item.type === 'adjustment' && item.adjustment_direction === 'in') return item.amount;
  if (item.type === 'adjustment' && item.adjustment_direction === 'out') return -item.amount;
  return 0;
}

function summarize(rows: Movement[], categories: Category[], year: number, month: number, budget: Budget | null): MonthlySummary {
  const categoryMap = new Map(categories.map((item) => [item.id, item]));
  const categoryTotals = new Map<number, number>();
  const dayTotals = new Map<string, number>();
  const expenses: Movement[] = [];
  let totalIncome = 0;
  let totalExpense = 0;
  let balance = 0;
  let necessary = 0;
  let unnecessary = 0;

  rows.forEach((item) => {
    balance += monthlyBalanceDelta(item);
    if (item.type === 'income') totalIncome += item.amount;
    if (item.type !== 'expense' && item.type !== 'purchase') return;
    expenses.push(item);
    totalExpense += item.amount;
    categoryTotals.set(item.category_id, (categoryTotals.get(item.category_id) || 0) + item.amount);
    dayTotals.set(item.date, (dayTotals.get(item.date) || 0) + item.amount);
    if (item.is_necessary) necessary += item.amount;
    else unnecessary += item.amount;
  });

  const categoryExpenses = [...categoryTotals.entries()].map(([categoryId, amount]) => ({
    category_id: categoryId,
    category: categoryMap.get(categoryId)?.name || 'Sin categoria',
    color: categoryMap.get(categoryId)?.color || '#6B7280',
    amount
  })).sort((a, b) => b.amount - a.amount);
  const highestDay = [...dayTotals.entries()].map(([date, amount]) => ({ date, amount })).sort((a, b) => b.amount - a.amount)[0];
  const messages: string[] = [];
  if (!rows.length) messages.push('No existen movimientos registrados en este periodo.');
  if (!budget) messages.push('No se ha definido presupuesto ni meta de ahorro para este mes.');

  return {
    year,
    month,
    total_income: totalIncome,
    total_expense: totalExpense,
    balance,
    saving_amount: balance,
    saving_rate: totalIncome ? (balance / totalIncome) * 100 : null,
    expense_rate: totalIncome ? (totalExpense / totalIncome) * 100 : null,
    average_daily_expense: dayTotals.size ? totalExpense / dayTotals.size : 0,
    highest_expense_day: highestDay,
    top_expense_category: categoryExpenses[0],
    necessary_expenses: necessary,
    unnecessary_expenses: unnecessary,
    category_expenses: categoryExpenses,
    top_expenses: expenses.sort((a, b) => b.amount - a.amount).slice(0, 10).map((item) => ({ id: item.id, date: item.date, description: item.description, amount: item.amount, category_id: item.category_id })),
    budget,
    messages
  };
}

function variation(current: number, previous: number) {
  return { absolute: current - previous, percent: previous ? ((current - previous) / previous) * 100 : null };
}

function compare(current: MonthlySummary, previous: MonthlySummary) {
  if (!previous.total_income && !previous.total_expense) return { current, previous, message: 'No existe mes anterior para comparar.', expense_variation: null, income_variation: null, balance_variation: null, largest_category_increase: null };
  const previousCategories = new Map(previous.category_expenses.map((item) => [item.category_id, item.amount]));
  const largest = current.category_expenses
    .map((item) => ({ ...item, increase: item.amount - (previousCategories.get(item.category_id) || 0) }))
    .filter((item) => item.increase > 0)
    .sort((a, b) => b.increase - a.increase)[0] || null;
  return {
    current,
    previous,
    message: null,
    expense_variation: variation(current.total_expense, previous.total_expense),
    income_variation: variation(current.total_income, previous.total_income),
    balance_variation: variation(current.balance, previous.balance),
    largest_category_increase: largest
  };
}

function recommendations(summary: MonthlySummary, comparison: ReturnType<typeof compare>): Recommendation[] {
  const result: Recommendation[] = [];
  const top = summary.top_expense_category;
  if (summary.total_expense > summary.total_income && top) result.push({ title: 'Gastos por encima de ingresos', explanation: `Este mes los gastos superaron los ingresos por ${money(summary.total_expense - summary.total_income)}.`, affected_value: summary.total_expense - summary.total_income, suggested_action: `Revisa primero la categoria ${top.category}.`, level: 'critico' });
  if (top && summary.total_expense && top.amount / summary.total_expense > 0.3) result.push({ title: 'Gasto concentrado', explanation: `${top.category} representa ${((top.amount / summary.total_expense) * 100).toFixed(1)}% del gasto mensual.`, affected_value: top.amount, suggested_action: 'Define un limite semanal para esta categoria.', level: 'advertencia' });
  if (comparison.expense_variation?.percent != null && comparison.expense_variation.absolute > 0) result.push({ title: 'Aumento frente al mes anterior', explanation: `Los gastos aumentaron ${comparison.expense_variation.percent.toFixed(1)}%.`, affected_value: comparison.expense_variation.absolute, suggested_action: 'Compara los movimientos que explican el aumento.', level: 'advertencia' });
  if (summary.budget?.saving_goal) {
    const difference = summary.saving_amount - summary.budget.saving_goal;
    result.push({ title: difference >= 0 ? 'Meta de ahorro cumplida' : 'Meta de ahorro pendiente', explanation: difference >= 0 ? `Superaste la meta por ${money(difference)}.` : `Faltan ${money(Math.abs(difference))}.`, affected_value: Math.abs(difference), suggested_action: difference >= 0 ? 'Conserva el excedente.' : 'Separa el ahorro antes de gastar.', level: difference >= 0 ? 'positivo' : 'advertencia' });
  }
  if (!result.length && summary.total_income) result.push({ title: 'Mes bajo control', explanation: `El balance mensual es ${money(summary.balance)}.`, affected_value: summary.balance, suggested_action: 'Continua registrando cada movimiento.', level: 'positivo' });
  return result;
}

export async function loadFinancialDataset({ uid, accountId, year, month, trendMonths, categories, budget, currentBalance }: { uid: string; accountId: number; year: number; month: number; trendMonths: number; categories: Category[]; budget: Budget | null; currentBalance: number }): Promise<FinancialDataset> {
  const trendStart = shiftMonth(year, month, -(Math.max(trendMonths, 2) - 1));
  const previous = shiftMonth(year, month, -1);
  const start = trendStart.year < previous.year || (trendStart.year === previous.year && trendStart.month < previous.month) ? trendStart : previous;
  return perfMeasureAsync('firestore-financial-dataset', async () => {
    const [movements, latestMovements] = await Promise.all([
      movementsService.getAllByRange({ uid, accountId, startDate: monthStart(start.year, start.month), endDate: monthEnd(year, month) }),
      movementsService.getLatestMovements({ uid, accountId, count: 5 })
    ]);
    return { movements, latestMovements, categories, budget, currentBalance };
  });
}

export function calculateFinancialAnalysis({ dataset, year, month, trendMonths }: { dataset: FinancialDataset; year: number; month: number; trendMonths: number }): FinancialAnalysis {
  performance.mark('financial-analysis:start');
  const groups = new Map<string, Movement[]>();
  dataset.movements.forEach((item) => {
    const key = item.date.slice(0, 7);
    const group = groups.get(key) || [];
    group.push(item);
    groups.set(key, group);
  });
  const previousPeriod = shiftMonth(year, month, -1);
  const current = summarize(groups.get(periodKey(year, month)) || [], dataset.categories, year, month, dataset.budget);
  const previous = summarize(groups.get(periodKey(previousPeriod.year, previousPeriod.month)) || [], dataset.categories, previousPeriod.year, previousPeriod.month, null);
  const comparison = compare(current, previous);
  const trends = Array.from({ length: trendMonths }, (_, index) => shiftMonth(year, month, index - trendMonths + 1)).map((period) => {
    const summary = summarize(groups.get(periodKey(period.year, period.month)) || [], dataset.categories, period.year, period.month, period.year === year && period.month === month ? dataset.budget : null);
    return { year: period.year, month: period.month, label: periodKey(period.year, period.month), income: summary.total_income, expense: summary.total_expense, saving: summary.saving_amount, balance: summary.balance };
  });
  performance.mark('financial-analysis:end');
  performance.measure('financial-analysis', 'financial-analysis:start', 'financial-analysis:end');
  return {
    summary: current,
    comparison,
    recommendations: recommendations(current, comparison),
    currentBalance: dataset.currentBalance,
    trends,
    latestMovements: decorateMovements(dataset.latestMovements, dataset.categories),
    documentsReadMaximum: 2000 + 20 + dataset.categories.length + 2
  };
}

export const analyticsService = {
  async load(uid: string, accountId: number, year: number, month: number, trendMonths: number, categories: Category[], budget: Budget | null, currentBalance: number) {
    const dataset = await loadFinancialDataset({ uid, accountId, year, month, trendMonths, categories, budget, currentBalance });
    return perfMeasureAsync('financial-analysis-calculation', async () => calculateFinancialAnalysis({ dataset, year, month, trendMonths }));
  }
};

File: backend/app/analytics.py
from calendar import monthrange
from datetime import date

from sqlmodel import Session, select

from .models import Budget, Category, CategoryBudget, Movement
from .schemas import Recommendation

EXPENSE_TYPES = {"expense", "purchase", "adjustment"}


def money(value: float) -> str:
    return "$" + f"{round(value):,}".replace(",", ".")


def previous_month(year: int, month: int) -> tuple[int, int]:
    return (year - 1, 12) if month == 1 else (year, month - 1)


def movements_for_month(session: Session, year: int, month: int) -> list[Movement]:
    rows = session.exec(select(Movement)).all()
    return [m for m in rows if m.date.year == year and m.date.month == month]


def income_total(rows: list[Movement]) -> float:
    return sum(m.amount for m in rows if m.type == "income")


def expense_total(rows: list[Movement]) -> float:
    return sum(m.amount for m in rows if m.type in EXPENSE_TYPES)


def category_expenses(session: Session, rows: list[Movement]) -> list[dict]:
    totals: dict[int, float] = {}
    for movement in rows:
        if movement.type in EXPENSE_TYPES:
            totals[movement.category_id] = totals.get(movement.category_id, 0) + movement.amount
    result = []
    for category_id, amount in sorted(totals.items(), key=lambda item: item[1], reverse=True):
        category = session.get(Category, category_id)
        result.append(
            {
                "category_id": category_id,
                "category": category.name if category else "Sin categoria",
                "color": category.color if category else "#6B7280",
                "amount": amount,
            }
        )
    return result


def top_category(session: Session, rows: list[Movement]) -> dict | None:
    categories = category_expenses(session, rows)
    return categories[0] if categories else None


def monthly_summary(session: Session, year: int, month: int) -> dict:
    rows = movements_for_month(session, year, month)
    total_income = income_total(rows)
    total_expense = expense_total(rows)
    balance = total_income - total_expense
    expense_days = {m.date for m in rows if m.type in EXPENSE_TYPES}
    by_category = category_expenses(session, rows)
    top = by_category[0] if by_category else None
    necessary = sum(m.amount for m in rows if m.type in EXPENSE_TYPES and m.is_necessary)
    unnecessary = sum(m.amount for m in rows if m.type in EXPENSE_TYPES and not m.is_necessary)
    top_expenses = sorted([m for m in rows if m.type in EXPENSE_TYPES], key=lambda m: m.amount, reverse=True)[:10]
    budget = session.exec(select(Budget).where(Budget.year == year, Budget.month == month)).first()

    return {
        "year": year,
        "month": month,
        "total_income": total_income,
        "total_expense": total_expense,
        "balance": balance,
        "saving_amount": balance,
        "saving_rate": (balance / total_income * 100) if total_income else None,
        "expense_rate": (total_expense / total_income * 100) if total_income else None,
        "average_daily_expense": (total_expense / len(expense_days)) if expense_days else 0,
        "highest_expense_day": highest_expense_day(rows),
        "top_expense_category": top,
        "necessary_expenses": necessary,
        "unnecessary_expenses": unnecessary,
        "category_expenses": by_category,
        "top_expenses": [
            {
                "id": m.id,
                "date": m.date.isoformat(),
                "description": m.description,
                "amount": m.amount,
                "category_id": m.category_id,
            }
            for m in top_expenses
        ],
        "budget": budget.model_dump() if budget else None,
        "messages": empty_messages(rows, budget),
    }


def highest_expense_day(rows: list[Movement]) -> dict | None:
    totals: dict[date, float] = {}
    for movement in rows:
        if movement.type in EXPENSE_TYPES:
            totals[movement.date] = totals.get(movement.date, 0) + movement.amount
    if not totals:
        return None
    day, amount = max(totals.items(), key=lambda item: item[1])
    return {"date": day.isoformat(), "amount": amount}


def empty_messages(rows: list[Movement], budget: Budget | None) -> list[str]:
    messages = []
    if not rows:
        messages.append("No existen movimientos registrados en este periodo.")
    if not budget:
        messages.append("No se ha definido presupuesto para este mes.")
        messages.append("No se ha definido meta de ahorro.")
    elif not budget.saving_goal:
        messages.append("No se ha definido meta de ahorro.")
    return messages


def comparison(session: Session, year: int, month: int) -> dict:
    current = monthly_summary(session, year, month)
    prev_year, prev_month = previous_month(year, month)
    previous_rows = movements_for_month(session, prev_year, prev_month)
    previous = monthly_summary(session, prev_year, prev_month)
    if not previous_rows:
        return {
            "current": current,
            "previous": previous,
            "message": "No existe mes anterior para comparar.",
            "expense_variation": None,
            "income_variation": None,
            "largest_category_increase": None,
        }
    return {
        "current": current,
        "previous": previous,
        "message": None,
        "expense_variation": variation(current["total_expense"], previous["total_expense"]),
        "income_variation": variation(current["total_income"], previous["total_income"]),
        "balance_variation": variation(current["balance"], previous["balance"]),
        "largest_category_increase": largest_category_increase(session, year, month, prev_year, prev_month),
    }


def variation(current: float, previous: float) -> dict:
    absolute = current - previous
    percent = (absolute / previous * 100) if previous else None
    return {"absolute": absolute, "percent": percent}


def largest_category_increase(session: Session, year: int, month: int, prev_year: int, prev_month: int) -> dict | None:
    current = {x["category_id"]: x for x in category_expenses(session, movements_for_month(session, year, month))}
    previous = {x["category_id"]: x for x in category_expenses(session, movements_for_month(session, prev_year, prev_month))}
    increases = []
    for category_id, item in current.items():
        diff = item["amount"] - previous.get(category_id, {"amount": 0})["amount"]
        increases.append({**item, "increase": diff})
    positive = [x for x in increases if x["increase"] > 0]
    return max(positive, key=lambda item: item["increase"]) if positive else None


def trends(session: Session, months: int = 6) -> list[dict]:
    today = date.today()
    result = []
    year, month = today.year, today.month
    for _ in range(max(1, min(months, 24))):
        summary = monthly_summary(session, year, month)
        result.append(
            {
                "year": year,
                "month": month,
                "label": f"{year}-{month:02d}",
                "income": summary["total_income"],
                "expense": summary["total_expense"],
                "saving": summary["saving_amount"],
                "balance": summary["balance"],
            }
        )
        year, month = previous_month(year, month)
    return list(reversed(result))


def recommendations(session: Session, year: int, month: int) -> list[Recommendation]:
    summary = monthly_summary(session, year, month)
    comp = comparison(session, year, month)
    recs: list[Recommendation] = []
    total_expense = summary["total_expense"]
    total_income = summary["total_income"]
    top = summary["top_expense_category"]

    if total_expense > total_income and top:
        recs.append(
            Recommendation(
                title="Gastos por encima de ingresos",
                explanation=f"Este mes tus gastos superaron tus ingresos por {money(total_expense - total_income)}. Revisa principalmente la categoria {top['category']}.",
                affected_value=total_expense - total_income,
                suggested_action="Reduce o aplaza gastos no necesarios antes de registrar nuevos compromisos.",
                level="critico",
            )
        )

    if top and total_expense:
        percent = top["amount"] / total_expense * 100
        if percent > 30:
            recs.append(
                Recommendation(
                    title="Alta concentracion por categoria",
                    explanation=f"La categoria {top['category']} representa el {percent:.1f}% de tus gastos del mes. Es tu mayor concentracion de consumo.",
                    affected_value=top["amount"],
                    suggested_action="Define un limite concreto para esta categoria y revisalo semanalmente.",
                    level="advertencia",
                )
            )

    expense_var = comp.get("expense_variation")
    largest = comp.get("largest_category_increase")
    if expense_var and expense_var["percent"] is not None:
        if expense_var["absolute"] > 0:
            category = largest["category"] if largest else "sin categoria identificada"
            recs.append(
                Recommendation(
                    title="Aumento frente al mes anterior",
                    explanation=f"Tus gastos aumentaron {expense_var['percent']:.1f}% frente al mes anterior. La mayor variacion estuvo en la categoria {category}.",
                    affected_value=expense_var["absolute"],
                    suggested_action="Compara los movimientos de esa categoria y decide cuales no se repiten.",
                    level="advertencia",
                )
            )
        elif expense_var["absolute"] < 0:
            recs.append(
                Recommendation(
                    title="Reduccion de gastos",
                    explanation=f"Tus gastos disminuyeron {abs(expense_var['percent']):.1f}% frente al mes anterior.",
                    affected_value=abs(expense_var["absolute"]),
                    suggested_action="Mantén las decisiones que redujeron el gasto este mes.",
                    level="positivo",
                )
            )

    budget = summary["budget"]
    if budget and budget.get("saving_goal"):
        diff = summary["saving_amount"] - budget["saving_goal"]
        if diff < 0:
            recs.append(
                Recommendation(
                    title="Meta de ahorro pendiente",
                    explanation=f"No alcanzaste tu meta de ahorro. Te faltaron {money(abs(diff))} para cumplirla.",
                    affected_value=abs(diff),
                    suggested_action="Separa primero el ahorro al recibir ingresos y ajusta gastos variables.",
                    level="advertencia",
                )
            )
        else:
            recs.append(
                Recommendation(
                    title="Meta de ahorro cumplida",
                    explanation=f"Cumpliste tu meta de ahorro. Superaste la meta por {money(diff)}.",
                    affected_value=diff,
                    suggested_action="Conserva este excedente como ahorro o fondo de emergencia.",
                    level="positivo",
                )
            )

    if budget and budget.get("unnecessary_expense_limit"):
        extra = summary["unnecessary_expenses"] - budget["unnecessary_expense_limit"]
        if extra > 0:
            recs.append(
                Recommendation(
                    title="Gastos no necesarios sobre el limite",
                    explanation=f"Tus gastos no necesarios superaron el limite definido por {money(extra)}.",
                    affected_value=extra,
                    suggested_action="Congela compras no necesarias hasta volver al limite.",
                    level="advertencia",
                )
            )

    small = [m for m in movements_for_month(session, year, month) if m.type in EXPENSE_TYPES and m.amount <= 30000]
    small_total = sum(m.amount for m in small)
    if len(small) >= 5:
        recs.append(
            Recommendation(
                title="Gastos pequenos acumulados",
                explanation=f"Detecte {len(small)} gastos pequenos que acumulados representan {money(small_total)}. Revisa si alguno puede reducirse.",
                affected_value=small_total,
                suggested_action="Agrupa compras pequeñas en un solo presupuesto semanal.",
                level="advertencia",
            )
        )

    if not recs and total_income:
        recs.append(
            Recommendation(
                title="Mes bajo control",
                explanation=f"Tu balance actual es {money(summary['balance'])} con ingresos reales por {money(total_income)}.",
                affected_value=summary["balance"],
                suggested_action="Continua registrando movimientos para mejorar el analisis.",
                level="positivo",
            )
        )
    return recs


def monthly_data(session: Session, year: int, month: int) -> dict:
    return {
        "summary": monthly_summary(session, year, month),
        "comparison": comparison(session, year, month),
        "recommendations": [r.model_dump() for r in recommendations(session, year, month)],
    }
